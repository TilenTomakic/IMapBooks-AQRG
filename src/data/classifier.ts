import { AnswerClass, DataService } from "./data";
import * as md5                     from "md5";
import * as fs                      from "fs-extra";
import * as serialize               from 'serialization';

function classifierFun() {
  const limdu = require('limdu');
  return new limdu.classifiers.NeuralNetwork();
}

export class ClassifierService {

  intentClassifier: any;

  classify(answer: AnswerClass) {
    return this.intentClassifier.classify(answer.toTrainVect());
  }

  async init(answersGroup: AnswerClass[]) {
    const id   = md5(answersGroup.map(x => x.rating + ':' + x.answer).join(';'));
    let file = './data/train/' + id + '.json';
    if (DataService.readyMode) {
      file = './data/trainReady/' + id + '.json';
    }

    if (fs.existsSync(file)) {
      console.log('Loading classifier ' + answersGroup[0].question + ' > ' + id);
      let limduClassifierSavedString = await fs.readFile(file);
      this.intentClassifier          = serialize.fromString(limduClassifierSavedString, __dirname);
    } else {
      console.log('Classifying ' + answersGroup[0].question + ' > ' + id);
      this.intentClassifier = classifierFun();
      this.intentClassifier.trainBatch(
        answersGroup.map(x => {
          return { input: x.toTrainVect(), output: x.rating }
        })
      );
      const limduClassifierExportString = serialize.toString(this.intentClassifier, classifierFun);
      await fs.writeFile(file, limduClassifierExportString);
    }
  }
}