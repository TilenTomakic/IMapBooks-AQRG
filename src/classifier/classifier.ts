import { AnswerClass, DataService } from "./data";
import * as md5                     from "md5";
import * as fs                      from "fs-extra";
import * as serialize               from 'serialization';
import * as puppeteer               from 'puppeteer';

function classifierFun() {
  const limdu = require('limdu');
  return new limdu.classifiers.NeuralNetwork();
}

export class ClassifierService {

  static readyMode = false;
  static trainMode = false;

  intentClassifier: any;

  constructor(public forModel: string) {
  }

  classify(answer: AnswerClass) {
    if (this.forModel === 'b') {
      return this.intentClassifier.classify(answer.toTrainVect());
    } else {
      return this.intentClassifier.classify(answer.toClassifyVectWithExtra());
    }
  }

  chrome() {
    // const browser = await puppeteer.launch();
    // const page = await browser.newPage();
    // await page.goto(`file:${path.join(__dirname, 'test.html')}`);
    // await page.screenshot({path: 'example.png'});
    // await browser.close();
  }

  async init(answersGroup: AnswerClass[]) {
    const id = md5(answersGroup.map(x => x.rating + ':' + x.answer).join(';'));
    let file = './data/train-' + this.forModel + '/';
    if (ClassifierService.readyMode || ClassifierService.trainMode) {
      file = './data/trainReady-' + this.forModel + '/';
    }
    await fs.ensureDir(file);
    file = file + id + '.json';

    if (fs.existsSync(file)) {
      // console.log('Loading classifier ' + answersGroup[0].question + ' > ' + id);
      let limduClassifierSavedString = await fs.readFile(file);
      this.intentClassifier          = serialize.fromString(limduClassifierSavedString, __dirname);
    } else {
      if (ClassifierService.readyMode) {
        throw new Error('ClassifierService is in ready mode but file was not found');
      }

      console.log('Classifying ' + file + ' > ' + answersGroup[ 0 ].question);
      this.intentClassifier = classifierFun();
      this.intentClassifier.trainBatch(
        answersGroup.map(x => {
          if (this.forModel === 'b') {
            return { input: x.toTrainVect(), output: x.rating }
          } else {
            return { input: x.toTrainVectWithExtra(), output: x.rating }
          }
        })
      );
      const limduClassifierExportString = serialize.toString(this.intentClassifier, classifierFun);
      await fs.writeFile(file, limduClassifierExportString);
    }
  }
}