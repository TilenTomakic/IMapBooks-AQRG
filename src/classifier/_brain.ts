import { AnswerClass, DataService } from "../data/data";
import * as md5                     from "md5";
import * as fs                      from "fs-extra";
import * as serialize               from 'serialization';
import * as puppeteer               from 'puppeteer';
import * as hash                    from 'object-hash';
import * as brain                   from 'brain.js';
import { ClassifierConst }          from "./const";
import { PredictResponse }          from "../server/interfaces";


export class BarinClassifierService {

  net = new brain.NeuralNetwork();
  // net = new brain.NeuralNetworkGPU();

  constructor(public forModel: string) {
  }

  classify(answer: AnswerClass) {
    let x;
    if (this.forModel === 'b') {
      x = this.net.run(answer.toTrainVect());
    } else {
      x = this.net.run(answer.toClassifyVectWithExtra());
    }
    return x;
  }

  classifyAsPredictResponse(answer: AnswerClass): PredictResponse {
    const x = this.classify(answer);
    const bestKey = Object.keys(x).sort((a, b) => x[b] - x[a])[0];
    return  {
      score      : + bestKey,
      probability: x[ bestKey ]
    }
  }

  async train(dataArray: { input: any, output: any }[]) {
    this.net.train(dataArray);

    /*
    const browser = await puppeteer.launch({
      headless: false
    });
    const page    = await browser.newPage();
    await page.setContent(`<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Training</title>
  <script src="https://cdn.rawgit.com/BrainJS/brain.js/master/browser.js"></script>
</head>
<body>
<script>
  const data = ${ JSON.stringify(dataArray) };
  const net = new brain.NeuralNetworkGPU();
  net.train(data);
  document.write(JSON.stringify(output));
</script>
</body>
</html>`, {
      timeout: 0
    });
    // await new Promise(r => {
    //   page.on('domcontentloaded' as any, () => {
    //   });
    // });
    const bodyHandle = await page.$('body');
    const html       = await page.evaluate(body => body.innerHTML, bodyHandle);
    await bodyHandle.dispose();
    await browser.close();*/
  }

  getData(answersGroup: AnswerClass[]) {
    return answersGroup.map(x => {
      if (this.forModel === 'b') {
        return { input: x.toTrainVect(), output: { [x.rating]: 1 } }
      } else {
        return { input: x.toTrainVectWithExtra(), output:  { [x.rating]: 1 }  }
      }
    })
  }

  async init(answersGroup: AnswerClass[]) {
    const data = this.getData(answersGroup);
    const id = hash(data, {
      unorderedObjects: true,
      unorderedArrays: true
    });
    let file = './data/brain/validation-' + this.forModel + '/';
    if (ClassifierConst.readyMode || ClassifierConst.trainMode) {
      file = './data/brain/server-' + this.forModel + '/';
    }
    await fs.ensureDir(file);
    file = file + id + '.json';

    if (fs.existsSync(file)) {
      console.log('BRAIN: Loading classifier ' + answersGroup[0].question + ' > ' + id);

      let netJson = await fs.readJSON(file);
      this.net.fromJSON(netJson);
    } else {
      console.log('BRAIN: Classifying ' + file + ' > ' + answersGroup[ 0 ].question);

      if (ClassifierConst.readyMode) {
        throw new Error('ClassifierService is in ready mode but file was not found');
      }

      await this.train(data);

      await fs.writeJSON(file, this.net.toJSON());
    }
  }
}