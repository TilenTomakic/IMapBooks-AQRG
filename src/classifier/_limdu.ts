import { AnswerClass, DataService } from "../data/data";
import * as fs                      from "fs-extra";
import * as serialize               from 'serialization';
import { ClassifierConst }          from "./const";
import { PredictResponse }          from "../server/interfaces";
import * as hash           from 'object-hash';

function classifierFun() {
  const limdu = require('limdu');
  return new limdu.classifiers.NeuralNetwork();
}

export class LimduClassifierService {

  intentClassifier: any;
  macroAverage: any;
  microAverage: any;

  constructor(public forModel: string) {
  }

  classify(answer: AnswerClass) {
    if (this.forModel === 'b') {
      return this.intentClassifier.classify(answer.toTrainVectors());
    } else {
      return this.intentClassifier.classify(answer.toClassifyVectorWithExtra());
    }
  }

  calcScore(cn) {
    let score = 0;
    if (cn[ 0 ] >= 0.25) {
      if (cn[ 0 ] < 0.75) {
        score = 0.5;
      } else {
        score = 1;
      }
    }
    return {
      score      : score,
      probability: cn[ 0 ]
    };
  }

  classifyAsPredictResponse(answer: AnswerClass): PredictResponse {
    const cn  = this.classify(answer);
    return this.calcScore(cn);
  }

  async init(answersGroup: AnswerClass[]) {


    const data = answersGroup.map(x => {
      if (this.forModel === 'b') {
        return { input: x.toTrainVectors(), output: x.rating }
      } else {
        return { input: x.toTrainVectorWithExtra(), output: x.rating }
      }
    });

    // await this.crossValidation(answersGroup, data);

    const id   = hash(data, {
      unorderedObjects: true,
      unorderedArrays : true
    });

    let file = './data/limdu/validation-' + this.forModel + '/';
    if (ClassifierConst.readyMode || ClassifierConst.trainMode) {
      file = './data/limdu/server-' + this.forModel + '/';
    }
    await fs.ensureDir(file);
    file = file + id + '.json';

    if (fs.existsSync(file)) {
      console.log('LIMDU: Loading saved classifier ' + file + ' > ' + answersGroup[ 0 ].question);

      let limduClassifierSavedString = await fs.readFile(file);
      this.intentClassifier          = serialize.fromString(limduClassifierSavedString, __dirname);
    } else {
      console.log('LIMDU: Classifying ' + file + ' > ' + answersGroup[ 0 ].question);
      if (ClassifierConst.readyMode) {
        throw new Error('ClassifierService is in ready mode but file was not found');
      }

      this.intentClassifier = classifierFun();
      this.intentClassifier.trainBatch(data);
      const limduClassifierExportString = serialize.toString(this.intentClassifier, classifierFun);
      await fs.writeFile(file, limduClassifierExportString);
    }
  }

  crossValidation(answersGroup: AnswerClass[], dataset) {
    if (ClassifierConst.readyMode) {
      return;
    }

    const limdu = require('limdu');
    let microAverage = new limdu.utils.PrecisionRecall();
    let macroAverage = new limdu.utils.PrecisionRecall();

    // Decide how many folds you want in your   k-fold cross-validation:
    const numOfFolds = 5;

    const self = this;
    return new Promise( resolve => {
      limdu.utils.partitions.partitions(dataset, numOfFolds, function(trainSet, testSet) {
        console.log("Training on "+trainSet.length+" samples, testing on "+testSet.length+" samples");
        let classifier = new limdu.classifiers.NeuralNetwork();
        const x = classifier.classify;
        classifier.classify = (z) => {
          const cn = x.call(classifier, z);
          const scr = self.calcScore(cn);
          // return { 0: scr.score > 0.5 ? 1 : 1 };
          return scr.score;
        };
        classifier.trainBatch(trainSet);
        limdu.utils.test(classifier, testSet, /* verbosity = */0,
          microAverage, macroAverage);
      });

      macroAverage.calculateMacroAverageStats(numOfFolds);
      //console.log("\n\nMACRO AVERAGE:"); console.dir(macroAverage.fullStats());
      this.macroAverage = macroAverage.fullStats();
      microAverage.calculateStats();
      //console.log("\n\nMICRO AVERAGE:"); console.dir(microAverage.fullStats());
      this.microAverage = microAverage.fullStats();

      console.log(`${ answersGroup[0].question }\nMACRO ${ macroAverage.shortStatsString }\nMICRO ${ microAverage.shortStatsString }\n\n`);
      fs.appendFileSync('./public/micro.txt', `${ this.forModel } -- ${ answersGroup[0].question }\nMACRO ${ macroAverage.shortStatsString }\nMICRO ${ microAverage.shortStatsString }\n\n`);

      resolve();
    });
  }
}
