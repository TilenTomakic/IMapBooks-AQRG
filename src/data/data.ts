import * as fs                                        from 'fs-extra';
import { DataSetInterface, normalizeOptions, Rating } from "./interfaces";
import * as natural                                   from 'natural';
import * as sw                                        from 'stopword';
import { PredictResponse }                            from "../server/interfaces";

(natural.PorterStemmer as any).attach();

const csv        = require('csvtojson');
const compromise = require('compromise');

class AnswerClass {
  question: string;
  answer: string;
  rating: 0 | 0.5 | 1;

  nlp: any;

  // FOR VALIDATION
  testA: PredictResponse;
  testB: PredictResponse;
  testC: PredictResponse;
  // <END>

  // FOR MODEL A
  similarity?: number;
  // <END>

  constructor(x: DataSetInterface) {
    this.question = x.Question;
    this.answer   = x.Response;
    this.rating   = ({
      [ Rating.The00 ]: 0,
      [ Rating.The05 ]: 0.5,
      [ Rating.The10 ]: 1,
    })[ x[ "Final.rating" ] ] as any;

    const nlp = this.nlp = compromise(x.Response);

    this.answer = x.Response;

    const norText       = nlp.normalize(normalizeOptions).out('array');
    const norTextNoStop = sw.removeStopwords(norText);
    // const tokens = normalize.tokenizeAndStem();
    // const tokensNoStops = normalizedNoStops.tokenizeAndStem();

    console.log();
  }

  // FOR MODEL A
  calcSimilarity(answer: string) {
    return this;
  }
  // <END>

  // FOR SUPER TEST
  toStats() {
    return {
      question: this.question,
      answer  : this.answer,
      rating  : this.rating,
      testA   : this.testA,
      testB   : this.testB,
      testC   : this.testC
    }
  }
  // <END>
}

export class DataService {

  story: string;
  dataSetString: string;
  rawData: DataSetInterface[];
  data: AnswerClass[];

  async init(withoutRow?: number) {
    this.story         = await fs.readFile('./data/Weightless.txt') + '';
    this.dataSetString = await fs.readFile('./data/Weightless_dataset_train.csv') + '';
    this.rawData       = await csv({
      delimiter: ',',
      flatKeys : true
    }).fromString(this.dataSetString);

    if (withoutRow !== undefined) {
      this.rawData.splice(withoutRow, 1);
    }

    this.data = this.rawData.map(x => new AnswerClass(x))
    return this;
  }
}
