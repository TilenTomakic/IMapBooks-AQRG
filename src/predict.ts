import { PredictRequest, PredictResponse } from "./server/interfaces";

export const predictService = new class {
  async predict(data: PredictRequest): Promise<PredictResponse> {
    return null;
  }
};
