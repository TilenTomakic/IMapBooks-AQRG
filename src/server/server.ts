import * as Fastify                        from 'fastify';
import { FastifyInstance }                 from 'fastify';
import * as path                           from "path";
import { predictService }                  from "../predict";

export const Server = new class {
  fastify: FastifyInstance;

  constructor() {
    const fastify = Fastify({
      logger: true
    });

    fastify.post('predict', async (request, reply) => {
      return predictService.predict(request.body);
    });

    fastify.register(require('fastify-static'), {
      root  : path.join(__dirname, 'public'),
      prefix: '/',
    });

    this.fastify = fastify;
  }

  listen() {
    this.fastify.listen(8080, (err, address) => {
      if (err) {
        throw err;
      }
      this.fastify.log.info(`server listening on ${address}`)
    });
  }
};

