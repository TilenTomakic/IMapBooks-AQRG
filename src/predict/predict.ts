import { PredictRequest, PredictResponse } from "../server/interfaces";
import { AnswerClass, DataService }        from "../data/data";
import { LimduClassifierService }          from "../classifier/_limdu";
import { BarinClassifierService }          from "../classifier/_brain";

export class PredictService {

  /**
   * Used for model A one answer per question.
   */
  fixedDataSet: { [ question: string ]: AnswerClass } = {};

  classifiers: { [ question: string ]: LimduClassifierService }          = {};
  classifiersWithExtra: { [ question: string ]: BarinClassifierService } = {};

  groups: { [ question: string ]: AnswerClass[] } = {};

  constructor(public dataService: DataService) {
  }

  async init() {
    this.dataService.dataA.forEach(x => {
      this.fixedDataSet[ x.question ] = this.fixedDataSet[ x.question ] || x;
      if (this.fixedDataSet[ x.question ].answerAdequacy() < x.answerAdequacy()) {
        this.fixedDataSet[ x.question ] = x;
      }
    });

    this.groups = this.dataService.data.reduce((a, c) => {
      a[ c.question ] = a[ c.question ] || [];
      a[ c.question ].push(c);
      return a;
    }, {});

    for (const g of Object.keys(this.groups)) {
      this.classifiers[ g ] = new LimduClassifierService('b');
      await this.classifiers[ g ].init(this.groups[ g ]);
    }

    for (const g of Object.keys(this.groups)) {
      this.classifiersWithExtra[ g ] = new BarinClassifierService('c');
      await this.classifiersWithExtra[ g ].init(this.groups[ g ]);
    }

    return this;
  }

  async predict(req: PredictRequest): Promise<PredictResponse> {
    return ({
      'A': this.predictA.bind(this),
      'B': this.predictB.bind(this),
      'C': this.predictC.bind(this)
    })[ req.modelId ](req);
  }

  async predictA(req: PredictRequest): Promise<PredictResponse> {
    const similarity = this.fixedDataSet[ req.question ].calcSimilarity(req.questionResponse);
    if (similarity < 0.05) {
      return {
        score      : 0,
        probability: similarity
      };
    }
    if (similarity < 0.1) {
      return {
        score      : 0.5,
        probability: similarity
      };
    }
    return {
      score      : 1,
      probability: similarity
    };
  }

  async predictB(req: PredictRequest): Promise<PredictResponse> {
    const other = AnswerClass.createFromPredictRequest(req);
    return this.classifiers[ other.question ].classifyAsPredictResponse(other);
  }

  async predictC(req: PredictRequest): Promise<PredictResponse> {
    const other = AnswerClass.createFromPredictRequest(req);
    return this.classifiersWithExtra[ other.question ].classifyAsPredictResponse(other);
  }
}

// let data = this.dataService.data
//   .filter(x => x.question === req.question)
//   .map((x, i) => ({ similarity: x.calcSimilarity(req.questionResponse), i, x }))
//   .sort((a, b) => b.similarity - a.similarity);
// if (data[ 0 ].similarity < 0.06) {
//   return {
//     score      : 0,
//     probability: data[ 0 ].similarity
//   };
// }
// return {
//   score      : data[ 0 ].x.rating,
//   probability: data[ 0 ].similarity
// };
/* const cn    = this.classifiersWithExtra[ other.question ].classify(other);
 let score   = 0;
 if (cn[ 0 ] < 0.25) {
 score = 0;
 } else {
 if (cn[ 0 ] < 0.75) {
 score = 0.5;
 } else {
 score = 1;
 }
 }

 return {
 score      : score,
 probability: cn[ 0 ]
 };*/