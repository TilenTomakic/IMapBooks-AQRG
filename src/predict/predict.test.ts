import { PredictService }    from "./predict";
import { DataService }       from "../data/data";
import { Server }            from "../server/server";
import { ClassifierService } from "../data/classifier";

DataService.readyMode = true;
ClassifierService.readyMode = true;

const dataService = new DataService();
let predictService: PredictService;

beforeAll(() => {
  return dataService
    .init()
    .then(dataService => {
      return (new PredictService(dataService)).init();
    })
    .then(ps => {
      predictService = ps;
      const server = new Server(ps);
      server.listen();
    })
});

describe('Predict test', () => {
  it('A model', async (done) => {
    const res = await predictService.predict({
      "modelId": "A",
      "question": "How does Shiranna feel as the shuttle is taking off?",
      "questionResponse": "She is nervous and excited."
    });
    expect(res.score).toMatchSnapshot();

    done();
  });

  it('B model', async (done) => {
    const res = await predictService.predict({
      "modelId": "B",
      "question": "How does Shiranna feel as the shuttle is taking off?",
      "questionResponse": "She is nervous and excited."
    });
    expect(res.score).toMatchSnapshot();

    done();
  });

  it('C model', async (done) => {
    const res = await predictService.predict({
      "modelId": "C",
      "question": "How does Shiranna feel as the shuttle is taking off?",
      "questionResponse": "She is nervous and excited."
    });
    expect(res.score).toBe(1);

    done();
  });
});


