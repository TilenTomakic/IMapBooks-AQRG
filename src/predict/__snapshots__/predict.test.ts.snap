// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Predict test A model 1`] = `0`;

exports[`Predict test B model 1`] = `0.5`;
